### PR TITLE
Backend

### PR DESCRIPTION
**Pull request by CatalinB7**
There are 2 directories, backend and frontend, but also 3 older files inside dogs_labeling directory that should have been removed.

### CODE REVIEW

File: dogs_labeling/frontend/src/picsManipulation.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        }
+        alreadySent[text][idx] = true;
+        let link = document.querySelectorAll("img")[idx].src;
+        await postPreferences({ id: getSessionId(), link, category: text });
+    }
+}
+
+function removeImage(idx: number) {
+    return async function deleteImg() {
+        let text = document.querySelector("h1").innerHTML;
```
Comment (CalinaCristian) [inline dogs_labeling/frontend/src/picsManipulation.ts:27]: Why do you need to check this stuff since you can have multiple h1's in the application. You should have (query param) routes as it was a requirement, and use that to determine on which of the pages you are on.
Also, since this logic is repeated, maybe make a function that determines on which view/route you are on.

File: dogs_labeling/frontend/src/picsManipulation.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    for (let i = 0; i < noPics; i++) {
+        insertNewCard(i, "block");
+    }
+}
+
+export function myClosure(idx: number) {
+    return async function clickedSavePicButton() {
+        let text = this.innerText;
+        text.includes("Silly") ? text = "silly" : text = "adorable";
+        if (alreadySent[text][idx]) {
```
[the remark this answers is not in the document]
Comment (CalinaCristian) [reply dogs_labeling/frontend/src/picsManipulation.ts:15]: Also, it's a pretty bad practice to keep the state of pictures that are already sent to the backend on the frontend code. The backend should be the source of truth (if you ever want to persist the session and the already loaded pics on refresh instead of requesting new random ones, this logic of yours would not work).

File: dogs_labeling/frontend/package.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    "ts-loader": "^9.2.3",
+    "typescript": "^4.3.5",
+    "webpack": "^5.0.0",
+    "webpack-cli": "^4.7.2"
+  },
+  "devDependencies": {
+    "css-loader": "^5.2.7",
+    "html-webpack-plugin": "^5.3.2",
+    "webpack-dev-server": "^3.11.2"
+  }
```
Comment (alexdudescu) [inline dogs_labeling/frontend/package.json:26]: When installing a package using npm you would use `npm i awesome-package` and that will add it to the dependencies but you can also add the `-D` or `--save-dev` flag and that will add it to the dev dependencies. When adding a new package to your project you should ask yourself if it's something you need for development only like the typescript compiler, typescript's @type packages or webpack, or if it's something that provides functionality for the consumer of your project and should end up in the actual build like a component library, lodash, etc.

File: dogs_labeling/frontend/src/picsManipulation.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    }
+}
+
+function removeImage(idx: number) {
+    return async function deleteImg() {
+        let text = document.querySelector("h1").innerHTML;
+        text = text.includes("silly") ? "silly" : "adorable";
+        let link = document.querySelectorAll("img")[idx].src;
+        await deletePreferences({ id: getSessionId(), category: text, link });
+        setReceivedLinks(getReceivedLinks().filter(elem => elem != link));
```
Comment (CalinaCristian) [inline dogs_labeling/frontend/src/picsManipulation.ts:31]: this is still optimistic ui, you don't call the endpoint do validate that the resource was actually deleted. You should call the get again here, and update the view based on what the actual data is in the backend.

File: dogs_labeling/index.html [hunk 5]
```diff
@@ -10,21 +10,20 @@
...
-
-
-
-<!--
-    js file loads at the end but i need the pics from beginning
-
-
- -->
\ No newline at end of file
+<script src="./script.js"></script>
```
Comment (CalinaCristian) [inline dogs_labeling/index.html:28]: Remove this outdatred file.

File: dogs_labeling/script.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
+const noPics = 8;
```
Comment (CalinaCristian) [inline dogs_labeling/script.js:1]: isn't this file outdated? Shouldn't this be removed?

File: dogs_labeling/frontend/index.html [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            </div>
+        </div>
+        <div class="pageContainer">
+            <button class="pageNav">⬅</button>
+            <button class="pageNav">➡</button>
+        </div>
+        
+    </div>
+
+    <script src="./dist/bundle.js"></script>
```
Comment (CalinaCristian) [inline dogs_labeling/frontend/index.html:39]: I don't like the position of the script here, you should put it in the head, and use [async/defer](https://www.digitalocean.com/community/tutorials/html-defer-async) attribute

File: dogs_labeling/frontend/src/script.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+document.querySelector('.form-btn').addEventListener("click", clickedStartButton); //send Name
+(document.querySelector("form") as HTMLFormElement).addEventListener("submit", clickedStartButton); //send Name
+document.querySelector('.refresh-btn').addEventListener("click", clickedRefreshButton) // Get other dogs
+document.querySelectorAll('.request-btn').forEach(btn => {
+    btn.addEventListener("click", clickedRequestButton); //get labeled dogs
+});
+document.querySelectorAll('.pageNav').forEach(elem => {
+    elem.addEventListener("click", changePage); //page navigation
+});
```
Comment (CalinaCristian) [inline dogs_labeling/frontend/src/script.ts:15]: why don't you use the onclick binding directly on the buttons?

File: dogs_labeling/frontend/src/httpServices.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    let response = await fetch('http://localhost:8080/preferences?' + new URLSearchParams(params as any), {
+        method: "GET",
+        headers: { 'Content-Type': 'application/json' },
+    });
+    let responseJSON: myResponse = await response.json();
+    return responseJSON;
+}
+
+export async function postPreferences(body: any) {
+    let response = await fetch('http://localhost:8080/preferences', {
```
Comment (alexdudescu) [inline dogs_labeling/frontend/src/httpServices.ts:15]:
```suggestion
    const baseUrl = 'http://localhost:8080/preferences';
    let response = await fetch(baseUrl, {
```

If for example the port your app is running on changes then you would have to go through all the instances where you use the hardcoded url and change it. In general, when you're using a value multiple times, think if would be useful to extract it as a const.

File: dogs_labeling/frontend/src/script.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+
+async function clickedStartButton() {
+    event.preventDefault();
```
Comment (CalinaCristian) [inline dogs_labeling/frontend/src/script.ts:19]: what event? you don't use it as parameter.

File: dogs_labeling/frontend/src/script.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        id: getSessionId(),
+        category: text,
+    });
+    setReceivedLinks(response.links);
+    makeElementsVisible(response, text);
+    setCurrentPage(1);
+    toggleNavButton();
+}
+
+function makeElementsVisible(response: myResponse, text: string) {
```
Comment (CalinaCristian) [inline dogs_labeling/frontend/src/script.ts:56]: this function seems a bit too complex and hard to follow, maybe you should refactor a bit the cards creation so this wouldn't be so complicated.

**Review by CalinaCristian — CHANGES_REQUESTED** [review]
Good job with this project! I've looked through the code, and I've left some suggestions, and I'd also add some keynotes here:

👍  I like the decoupling of backend and frontend and the usage of webpack to bundle files & dev server !
👍  I also liked that you documented the functions (at least on backend)
👍  I really like the backend code structure, it is very easy to follow !

👎  The page doesn't seem to be responsive (except for the usage of flex)
👎  You don't seem to use routings, which was one of the requirements, and you use some h1 query weird logic to determine on which page you are on (also you should be able to go to all the rouites, including the user selection..so maybe add a logout button?)
👎 The FE code still seems a bit cluttered, and you keep a lot of logic on FE instead of relying on the BE to give you the actual data. If there will ever be a requirement
to have multiple users with access to the same adorable/silly lists, there will need to be a lot of changes to the code.

**Review by alexdudescu — COMMENTED** [review]
I really like the frontend dev setup with webpack. 

Also I'm in alignment with what Cristi mentioned. The logic on the fe side also feels a bit "backend-esque". You could probably take advantage more of language features such as the forEach, map, reduce functions that would make your code cleaner. 

Overall nice project! ♥

File: dogs_labeling/frontend/package.json (hunk 3 above)
Comment (CalinaCristian) [reply dogs_labeling/frontend/package.json:26]: i like how you tagged @type which is an actual github account

File: dogs_labeling/frontend/src/script.ts (hunk 8 above)
Comment (CatalinB7) [reply dogs_labeling/frontend/src/script.ts:34]: like this?
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-3cb787b8ed440dcf7e44c103d614a482c34866350f31e8b770afc564cab9225cR31-R34

File: dogs_labeling/frontend/src/script.ts (hunk 10 above)
Comment (CatalinB7) [reply dogs_labeling/frontend/src/script.ts:38]: passed event as parameter now
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-3cb787b8ed440dcf7e44c103d614a482c34866350f31e8b770afc564cab9225cR36-R38

File: dogs_labeling/frontend/package.json (hunk 3 above)
Comment (CatalinB7) [reply dogs_labeling/frontend/package.json]: moved most of the dependencies to devDependencies
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-e575b5aad16756346f1e1ab58b0aa1d67c54b09fefcf17c575ce413776dd7f40R1-R22

File: dogs_labeling/frontend/src/httpServices.ts (hunk 9 above)
Comment (CatalinB7) [reply dogs_labeling/frontend/src/httpServices.ts]: followed your advice and i also created a generic fetch function
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-b0abc765a3b7c7158a09dc4d11771c0cc58bf9a6d6f158c2791d2ad656fbb37fR4-R19

File: dogs_labeling/backend/src/db.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
+type Category = ("silly" | "adorable");
+type Dict = { [key in Category]: string[] };
```
[the remark this answers is not in the document]
Comment (CatalinB7) [reply dogs_labeling/backend/src/db.ts]: ```type categoriesToLinks = { [key in Category]: string[] };```

File: dogs_labeling/frontend/src/styles.css [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+        transform: rotate(-30deg) translateY(105px);
+    }
+      100% {
+        transform-origin: center 105px;
+        transform: rotate(0deg) translateY(105px);
+    }
+  }
+
+  .trashcan:hover {
+    animation: rotation 3s;
```
[the remark this answers is not in the document]
Comment (CatalinB7) [reply dogs_labeling/frontend/src/styles.css]: on hover changing only color
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-86bd6ef696b30ab1e82966387c57901d489f42614d8fa8f4797095050f58a6edR148-R151

File: dogs_labeling/frontend/src/appState.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
+//probably i should manage the state in another way
```
[the remark this answers is not in the document]
Comment (CatalinB7) [reply dogs_labeling/frontend/src/appState.ts]: done as suggested
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-d9270f86a9087ab9516b7bc6abf59f12e83ef38e8899500a06fb41f48b68d173R22-R113

File: dogs_labeling/frontend/src/picsManipulation.ts (hunk 4 above)
Comment (CatalinB7) [reply dogs_labeling/frontend/src/picsManipulation.ts]: less optimistic more realistic :D
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-5835e833b0368ccea5f129aae6a43784e4c574154f58b8b560bd84d3ceb26ccaR33-R37

File: dogs_labeling/frontend/src/httpServices.ts [hunk 15]
```diff
@@ -0,0 +1,48 @@
+import { getSessionId, setSessionId } from "./appState";
+import { myResponse, fetchParams } from "./models";
+//should these fetching functions be more generic?
+//receive parameters for method and url maybe
```
[the remark this answers is not in the document]
Comment (CatalinB7) [reply dogs_labeling/frontend/src/httpServices.ts]: followed your advice and i created a generic fetch function
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-b0abc765a3b7c7158a09dc4d11771c0cc58bf9a6d6f158c2791d2ad656fbb37fR4-R19

File: dogs_labeling/frontend/src/models.ts [hunk 16]
```diff
@@ -0,0 +1,8 @@
+export interface myResponse {
```
[the remark this answers is not in the document]
Comment (CatalinB7) [reply dogs_labeling/frontend/src/models.ts]: done
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-ba7c7caf9a12bff712e797fb6a07161fcdb3706738bfb18fbb62b24aacdf202fR1-R11

File: dogs_labeling/frontend/src/picsManipulation.ts [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+import { noPics, alreadySent, resetAlreadySent, picWidth, getSessionId, setReceivedLinks, getReceivedLinks, getCurrentPage, toggleNavButton } from "./appState";
+
+export function createCards() {
+    //creating cards containing pics and 2 buttons
+    for (let i = 0; i < noPics; i++) {
+        insertNewCard(i, "block");
+    }
+}
+
+export function myClosure(idx: number) {
```
[the remark this answers is not in the document]
Comment (CatalinB7) [reply dogs_labeling/frontend/src/picsManipulation.ts]: renamed it ```export function savePic(idx: number)```

File: dogs_labeling/frontend/src/picsManipulation.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (CatalinB7) [reply dogs_labeling/frontend/src/picsManipulation.ts]: disabling button after it got clicked:
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-5835e833b0368ccea5f129aae6a43784e4c574154f58b8b560bd84d3ceb26ccaR13-R17

I keep the state of pictures on FE, but the backend also checks weather or not the request is valid.

File: dogs_labeling/frontend/src/picsManipulation.ts (hunk 1 above)
Comment (CatalinB7) [reply dogs_labeling/frontend/src/picsManipulation.ts]: Routes:
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-3cb787b8ed440dcf7e44c103d614a482c34866350f31e8b770afc564cab9225cR8-R23

Now checking the route in order to know what do display.

File: dogs_labeling/frontend/index.html (hunk 7 above)
Comment (CatalinB7) [reply dogs_labeling/frontend/index.html]: moved them
https://github.com/CatalinB7/dogs-labeling/pull/2/files#diff-4a3cfa61237c77c4d40e8b3f8324d29c6b86966ff024895baa9728ba0c45ccefR10-R12

File: dogs_labeling/script.js (hunk 6 above)
Comment (CatalinB7) [reply dogs_labeling/script.js]: done
